src/csharp/Grpc.Core/GrpcEnvironment.cs

```diff
@@ -196,8 +202,30 @@ public static ILogger Logger
         /// </summary>
         public static void SetLogger(ILogger customLogger)
         {
-            GrpcPreconditions.CheckNotNull(customLogger, "customLogger");
-            logger = customLogger;
+            lock (loggerLock)
+            {
+                GrpcPreconditions.CheckNotNull(customLogger, "customLogger");
+                logger = customLogger;
+                loggerCache.Clear();
+            }
+        }
+
+        internal static ILogger GetLoggerForType<T>() {
+            lock (loggerLock)
```

Comment:
It may well be worth using a ConcurrentDictionary - if we think that this is on a hot path, anyway.